app/models/mixins/assignment_mixin.rb

```diff
@@ -21,41 +21,74 @@ def all_assignments(tag = nil)
     cache_with_timeout(:assignments_cached, 1.minute) { assignments }
   end
 
+  def build_object_tag_path(obj, klass = nil)
+    if obj.kind_of?(ActiveRecord::Base) # obj is a CI
+      "#{obj.class.base_model.name.underscore}/id/#{obj.id}"
+    else                                # obj is the id of an instance of <klass>
+      raise _("Class must be specified when object is an integer") if klass.nil?
+      "#{klass.underscore}/id/#{obj}"
+    end
+  end
+
   def assign_to_objects(objects, klass = nil)
     # objects => A single item or array of items
     #   item  => A CI instance (not classification) or a CI id of klass
     # klass   => The class of the object that self is to be assigned to - (Takes both forms - Host or host, EmsCluster or ems_cluster)
     objects.to_miq_a.each do |obj|
-      if obj.kind_of?(ActiveRecord::Base) # obj is a CI
-        tag = "#{obj.class.base_model.name.underscore}/id/#{obj.id}"
-      else                                # obj is the id of an instance of <klass>
-        raise _("Class must be specified when object is an integer") if klass.nil?
-        tag = "#{klass.underscore}/id/#{obj}"
-      end
+      tag = build_object_tag_path(obj, klass)
       tag_add(tag, :ns => namespace)
     end
     reload
   end
 
+  def unassign_objects(objects, klass = nil)
+    # objects => A single item or array of items
+    #   item  => A CI instance (not classification) or a CI id of klass
+    # klass   => The class of the object that self is to be unassigned from - (Takes both forms - Host or host, EmsCluster or ems_cluster)
+    objects.to_miq_a.each do |obj|
+      tag = build_object_tag_path(obj, klass)
+      tag_remove(tag, :ns => namespace)
+    end
+    reload
+    get_assigned_tos
+  end
+
+  def build_tag_tagging_path(obj, klass)
+    unless obj.kind_of?(ActiveRecord::Base) # obj is the id of a classification entry instance
+      id = obj
+      obj = Classification.find_by(:id => id)
+      if obj.nil?
+        _log.warn("Unable to find classification with id [#{id}], skipping assignment")
+        nil
```

Comment:
Should this return nil here? It looks like the calling code is checking for a nil returned from this method and skipping.

Reply:
that's the idea - if we don't have a classification for a tag, we don't want to call `tag_add`, hence the skip.

Before I split this code into its own method, there was a `next` statement after the warning. Returning nil here and checking for nil in `assign_to_tags` ensures `assign_to_tags` will retain its previous behavior and will not allow creation of invalid assignments (from what I understand, a tag with no classification is not a valid tag for assignment)